composer/trainer/mosaic_fsdp_utils.py

```diff
@@ -717,3 +717,83 @@ def set_optimizer_state_dict(
 
             _verify_state_dict({}, optim_state_dict, info)
             _load_optim_state_dict(model, optimizers, optim_state_dict, info)
+
+
+    # torch2.3 patch to fix https://github.com/pytorch/pytorch/issues/125740
+    from torch.distributed.checkpoint.default_planner import (
+        create_default_global_save_plan,
+        DefaultSavePlanner,
+        _validate_global_plan,
+    )
+    import dataclasses
+    from collections import defaultdict, ChainMap
+    from typing import Dict, List, Set, TYPE_CHECKING
+
+    from torch.distributed.checkpoint.planner import SavePlan, WriteItem
+    from torch.distributed.checkpoint.metadata import MetadataIndex, Metadata
+
+    def dedup_save_plans(all_plans: List[SavePlan]) -> List[SavePlan]:
+        """
+        Removes duplicate entries from appearing on multiple SavePlans. For each duplicate across
+        a set of SavePlans, only the smallest SavePlan in terms of planned storage keeps the entry.
+        """
+
+        write_item_to_plan_indices: Dict[MetadataIndex, Set[int]] = defaultdict(set)
+        write_item_idx_to_write_item: Dict[MetadataIndex, WriteItem] = {}
+        for plan_idx, plan in enumerate(all_plans):
+            for write_item in plan.items:
+                # map each write item to its plan
+                write_item_to_plan_indices[write_item.index].add(plan_idx)
+                write_item_idx_to_write_item[write_item.index] = write_item
+
+        # put item in the plan with the smallest size and remove it from the other plan_indices
+        to_remove: List[Set] = [set() for _ in range(len(all_plans))]
+        plan_to_size = [0] * len(all_plans)
+        for write_item_idx, plan_indices in write_item_to_plan_indices.items():
+            # this line is the fix, to keep the duplicated tensors on the same rank
+            select_plan_idx = min(plan_indices, key=lambda plan_idx: plan_idx)
```

Comment:
Most of the code in this files are copied from torch. This line is the only change, to select the lowest rank, instead of selecting the rank with minimum storage.